### PR TITLE
Parallel version of calcError in statmodel

### PR DESCRIPTION
**Pull request by LaurentBerger**
Improvement (ParallelLoopBody) of calcError function in StatModel

### CODE REVIEW

File: modules/ml/src/inner_functions.cpp [hunk 1]
```diff
@@ -72,7 +72,67 @@ bool StatModel::train( InputArray samples, int layout, InputArray responses )
...
 }
 
-float StatModel::calcError( const Ptr<TrainData>& data, bool testerr, OutputArray _resp ) const
+class ParallelCalcError : public ParallelLoopBody
+{
+private:
+    const Ptr<TrainData>& data;
+    bool testerr;
+    Mat &resp;
+    bool verbose;
```
Comment (alalek) [inline modules/ml/src/inner_functions.cpp]: Not used, even in commented code

File: modules/ml/src/inner_functions.cpp [hunk 2]
```diff
@@ -84,37 +144,28 @@ float StatModel::calcError( const Ptr<TrainData>& data, bool testerr, OutputArra
...
-        if( !resp.empty() )
-            resp.at<float>(i) = val;
-        /*if( i < 100 )
-        {
-            printf("%d. ref %.1f vs pred %.1f\n", i, val0, val);
-        }*/
-    }
+    vector<double> errStrip(n);
+    for (int i = 0; i < errStrip.size(); i++)
+        errStrip[i] = 0;
```
Comment (alalek) [inline modules/ml/src/inner_functions.cpp]: Eliminate init loop:
```
vector<double> errStrip(n, 0.0);
```

File: modules/ml/src/inner_functions.cpp [hunk 3]
```diff
@@ -72,7 +72,67 @@ bool StatModel::train( InputArray samples, int layout, InputArray responses )
...
+    void Verbose(bool b) { verbose = b; }
+    virtual void operator()(const Range& range) const
+    {
+        int idxErr = range.start;
+        CV_TRACE_FUNCTION_SKIP_NESTED();
+        Mat samples = data->getSamples();
+        int layout = data->getLayout();
+        Mat sidx = testerr ? data->getTestSampleIdx() : data->getTrainSampleIdx();
+        const int* sidx_ptr = sidx.ptr<int>();
+        int i, n = (int)sidx.total();
```
Comment (alalek) [inline modules/ml/src/inner_functions.cpp]: "n" - used?
Declaration of "int i" is better to move into "for" statement.

File: modules/ml/src/inner_functions.cpp [hunk 4]
```diff
@@ -84,37 +144,28 @@ float StatModel::calcError( const Ptr<TrainData>& data, bool testerr, OutputArra
...
-        }*/
-    }
+    vector<double> errStrip(n);
+    for (int i = 0; i < errStrip.size(); i++)
+        errStrip[i] = 0;
+    ParallelCalcError x(data, testerr, resp, this,errStrip);
+
+    parallel_for_(Range(0,n),x);
+
+    for (int i = 0; i < errStrip.size(); i++)
```
Comment (alalek) [inline modules/ml/src/inner_functions.cpp]: int -> size_t

**Comment by LaurentBerger** [comment]
I don't understand test errors. I haven't got any errors in debug mode but I have got error only in release mode. in debug : 
```
CTEST_FULL_OUTPUT
OpenCV version: 3.3.1-dev
OpenCV VCS version: 3.3.1-129-g2bbd99c8a
Build type: debug
Parallel framework: ms-concurrency
CPU features: mmx sse sse2 sse3
Intel(R) IPP optimization: enabled
Intel(R) IPP version: ippIP AVX2 (l9) 2017.0.3 (-) Jul 31 2017
OpenCL is disabled
[==========] Running 37 tests from 14 test cases.
[----------] Global test environment set-up.
[----------] 1 test from ML_KMeans
[ RUN      ] ML_KMeans.accuracy
[       OK ] ML_KMeans.accuracy (293 ms)
[----------] 1 test from ML_KMeans (295 ms total)

[----------] 2 tests from ML_KNearest
[ RUN      ] ML_KNearest.accuracy
[       OK ] ML_KNearest.accuracy (92 ms)
[ RUN      ] ML_KNearest.save_load
[       OK ] ML_KNearest.save_load (620 ms)
[----------] 2 tests from ML_KNearest (713 ms total)

[----------] 3 tests from ML_EM
[ RUN      ] ML_EM.accuracy
[       OK ] ML_EM.accuracy (4210 ms)
[ RUN      ] ML_EM.save_load
[       OK ] ML_EM.save_load (17 ms)
[ RUN      ] ML_EM.classification
[       OK ] ML_EM.classification (4010 ms)
[----------] 3 tests from ML_EM (8239 ms total)

[----------] 2 tests from ML_LR
[ RUN      ] ML_LR.accuracy
[       OK ] ML_LR.accuracy (7554 ms)
[ RUN      ] ML_LR.save_load
[       OK ] ML_LR.save_load (7505 ms)
[----------] 2 tests from ML_LR (15060 ms total)

[----------] 3 tests from ML_DTree
[ RUN      ] ML_DTree.regression
[       OK ] ML_DTree.regression (3395 ms)
[ RUN      ] ML_DTree.save_load
[       OK ] ML_DTree.save_load (1119 ms)
[ RUN      ] ML_DTree.legacy_load
[       OK ] ML_DTree.legacy_load (93 ms)
[----------] 3 tests from ML_DTree (4609 ms total)

[----------] 3 tests from ML_Boost
[ RUN      ] ML_Boost.regression
[       OK ] ML_Boost.regression (13311 ms)
[ RUN      ] ML_Boost.save_load
[       OK ] ML_Boost.save_load (7967 ms)
[ RUN      ] ML_Boost.legacy_load
[       OK ] ML_Boost.legacy_load (175 ms)
[----------] 3 tests from ML_Boost (21455 ms total)

[----------] 4 tests from ML_RTrees
[ RUN      ] ML_RTrees.regression
[       OK ] ML_RTrees.regression (255323 ms)
[ RUN      ] ML_RTrees.getVotes
[       OK ] ML_RTrees.getVotes (4 ms)
[ RUN      ] ML_RTrees.save_load
[       OK ] ML_RTrees.save_load (35752 ms)
[ RUN      ] ML_RTrees.legacy_load
[       OK ] ML_RTrees.legacy_load (495 ms)
[----------] 4 tests from ML_RTrees (291575 ms total)

[----------] 1 test from ML_NBAYES
[ RUN      ] ML_NBAYES.regression_5911
[       OK ] ML_NBAYES.regression_5911 (2 ms)
[----------] 1 test from ML_NBAYES (3 ms total)

[----------] 1 test from ML_NaiveBayes
[ RUN      ] ML_NaiveBayes.save_load
[       OK ] ML_NaiveBayes.save_load (290 ms)
[----------] 1 test from ML_NaiveBayes (291 ms total)

[----------] 5 tests from ML_SVM
[ RUN      ] ML_SVM.save_load
[       OK ] ML_SVM.save_load (1412 ms)
[ RUN      ] ML_SVM.legacy_load
[       OK ] ML_SVM.legacy_load (48 ms)
[ RUN      ] ML_SVM.trainauto
[       OK ] ML_SVM.trainauto (47 ms)
[ RUN      ] ML_SVM.trainAuto_regression_5369
[       OK ] ML_SVM.trainAuto_regression_5369 (36 ms)
[ RUN      ] ML_SVM.getSupportVectors
[       OK ] ML_SVM.getSupportVectors (1 ms)
[----------] 5 tests from ML_SVM (1550 ms total)

[----------] 2 tests from ML_ANN
[ RUN      ] ML_ANN.save_load
[       OK ] ML_ANN.save_load (1417 ms)
[ RUN      ] ML_ANN.legacy_load
[       OK ] ML_ANN.legacy_load (53 ms)
[----------] 2 tests from ML_ANN (1471 ms total)

[----------] 1 test from MV_SVMSGD
[ RUN      ] MV_SVMSGD.save_load
[       OK ] MV_SVMSGD.save_load (209 ms)
[----------] 1 test from MV_SVMSGD (209 ms total)

[----------] 1 test from ML_NBayes
[ RUN      ] ML_NBayes.legacy_load
[       OK ] ML_NBayes.legacy_load (28 ms)
[----------] 1 test from ML_NBayes (30 ms total)

[----------] 8 tests from ML_SVMSGD
[ RUN      ] ML_SVMSGD.legacy_load
[       OK ] ML_SVMSGD.legacy_load (16 ms)
[ RUN      ] ML_SVMSGD.trainSameScale2
[       OK ] ML_SVMSGD.trainSameScale2 (1664 ms)
[ RUN      ] ML_SVMSGD.trainSameScale5
[       OK ] ML_SVMSGD.trainSameScale5 (1746 ms)
[ RUN      ] ML_SVMSGD.trainSameScale100
[       OK ] ML_SVMSGD.trainSameScale100 (3153 ms)
[ RUN      ] ML_SVMSGD.trainDifferentScales2
[       OK ] ML_SVMSGD.trainDifferentScales2 (1655 ms)
[ RUN      ] ML_SVMSGD.trainDifferentScales5
[       OK ] ML_SVMSGD.trainDifferentScales5 (1689 ms)
[ RUN      ] ML_SVMSGD.trainDifferentScales100
[       OK ] ML_SVMSGD.trainDifferentScales100 (3207 ms)
[ RUN      ] ML_SVMSGD.twoPoints
[       OK ] ML_SVMSGD.twoPoints (10 ms)
[----------] 8 tests from ML_SVMSGD (13148 ms total)

[----------] Global test environment tear-down
[==========] 37 tests from 14 test cases ran. (358662 ms total)
[  PASSED  ] 37 tests.

  YOU HAVE 3 DISABLED TESTS
```

In release I have got exactly same errors

**Pull request merged**